src/AspNet.Security.OAuth.VkId/VkIdAuthenticationHandler.cs

```diff
@@ -0,0 +1,337 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (http://www.apache.org/licenses/LICENSE-2.0)
+ * See https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers
+ * for more information concerning the license and the contributors participating to this project.
+ */
+
+using System.Globalization;
+using System.Net;
+using System.Net.Http.Headers;
+using System.Security.Claims;
+using System.Security.Cryptography;
+using System.Text;
+using System.Text.Encodings.Web;
+using System.Text.Json;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.WebUtilities;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.Extensions.Primitives;
+using Base64UrlEncoder = Microsoft.AspNetCore.Authentication.Base64UrlTextEncoder;
+
+namespace AspNet.Security.OAuth.VkId;
+
+public sealed partial class VkIdAuthenticationHandler : OAuthHandler<VkIdAuthenticationOptions>
+{
+    public VkIdAuthenticationHandler(
+        [NotNull] IOptionsMonitor<VkIdAuthenticationOptions> options,
+        [NotNull] ILoggerFactory logger,
+        [NotNull] UrlEncoder encoder)
+        : base(options, logger, encoder)
+    {
+    }
+
+    protected override string BuildChallengeUrl(
+        [NotNull] AuthenticationProperties properties,
+        [NotNull] string redirectUri)
+    {
+        var parameter = Options.Scope;
+        var scopes = FormatScope(parameter);
+
+        // It's mandatory to use PKCE
+        var data = RandomNumberGenerator.GetBytes(32);
+        var codeVerifierKey = Base64UrlEncoder.Encode(data);
+        properties.Items.Add(OAuthConstants.CodeVerifierKey, codeVerifierKey);
+
+        var query = new Dictionary<string, string?>
+        {
+            ["response_type"] = "code",
+            ["client_id"] = Options.ClientId,
+            ["scope"] = scopes,
+            ["redirect_uri"] = redirectUri,
+            ["state"] = Options.StateDataFormat.Protect(properties),
+            ["code_challenge"] = WebEncoders.Base64UrlEncode(SHA256.HashData(Encoding.UTF8.GetBytes(codeVerifierKey))),
+            ["code_challenge_method"] = OAuthConstants.CodeChallengeMethodS256
+        };
+        return QueryHelpers.AddQueryString(Options.AuthorizationEndpoint, query);
+    }
+
+    protected override async Task<HandleRequestResult> HandleRemoteAuthenticateAsync()
+    {
+        var query = Request.Query;
+        var properties = Options.StateDataFormat.Unprotect(query["state"]);
+        if (properties is null)
+        {
+            return HandleRequestResult.Fail("The oauth state was missing or invalid.");
+        }
+
+        // OAuth2 10.12 CSRF
+        if (ValidateCorrelationId(properties) is false)
```

Comment:
It's not technically wrong, but `!ValidateCorrelationId(properties)` would be easier to read (and consistent with the other providers).